### ISSUE TITLE
sched-simple allocates a node that is not online

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#4290]
Problem: 8 node cluster has  rank 0 excluded, and rank 4 offline (it does not exist but it is configured), however, rank 4 seems to have been allocated to a job by sched-simple even though it never appears in a resource.acquire "up" response.

This should have hung waiting for another node to come online:
```
$ flux mini run -N7 hostname
0.495s: job.exception type=prolog severity=0 prolog failed with status=17408
```

```
$ sudo flux dmesg
2022-04-19T01:42:34.473718Z broker.debug[0]: insmod connector-local
2022-04-19T01:42:34.473932Z broker.info[0]: start: none->join 52.3214ms
2022-04-19T01:42:34.474068Z broker.info[0]: parent-none: join->init 0.088333ms
2022-04-19T01:42:34.477071Z broker.info[0]: online:  (ranks )
2022-04-19T01:42:34.488180Z connector-local.debug[0]: allow-guest-user=true
2022-04-19T01:42:34.488217Z connector-local.debug[0]: allow-root-owner=true
2022-04-19T01:42:34.744635Z broker.debug[0]: insmod barrier
2022-04-19T01:42:34.880016Z broker.debug[0]: insmod content-sqlite
2022-04-19T01:42:34.926633Z broker.debug[0]: content backing store: enabled content-sqlite
2022-04-19T01:42:35.019088Z broker.debug[0]: insmod kvs
2022-04-19T01:42:35.021593Z kvs.info[0]: restored KVS from checkpoint on 2022-04-19T01:29:02Z
2022-04-19T01:42:35.097672Z broker.debug[0]: insmod kvs-watch
2022-04-19T01:42:35.203546Z broker.debug[0]: insmod resource
2022-04-19T01:42:35.307399Z resource.debug[0]: reslog_cb: resource-init event posted
2022-04-19T01:42:35.310529Z resource.debug[0]: reslog_cb: resource-define event posted
2022-04-19T01:42:35.360924Z broker.debug[0]: insmod cron
2022-04-19T01:42:35.361763Z cron.info[0]: synchronizing cron tasks to event heartbeat.pulse
2022-04-19T01:42:35.429630Z broker.debug[0]: insmod job-manager
2022-04-19T01:42:35.915093Z job-manager.info[0]: restart: 103 jobs
2022-04-19T01:42:35.915129Z job-manager.info[0]: restart: 0 running jobs
2022-04-19T01:42:35.916390Z job-manager.debug[0]: restart: max_jobid=787959464853504
2022-04-19T01:42:35.916530Z job-manager.debug[0]: duration-validator: updated expiration to 0.00
2022-04-19T01:42:35.978705Z broker.debug[0]: insmod job-info
2022-04-19T01:42:36.044940Z broker.debug[0]: insmod job-list
2022-04-19T01:42:36.240851Z job-list.debug[0]: job_state_init_from_kvs: read 103 jobs
2022-04-19T01:42:36.309965Z broker.debug[0]: insmod job-archive
2022-04-19T01:42:36.631900Z broker.debug[0]: insmod job-ingest
2022-04-19T01:42:36.632733Z job-ingest.debug[0]: configuring with plugins=(null), args=(null)
2022-04-19T01:42:36.635039Z job-ingest.debug[0]: fluid ts=46966045ms
2022-04-19T01:42:36.694404Z broker.debug[0]: insmod job-exec
2022-04-19T01:42:36.695287Z job-exec.debug[0]: using default shell path /usr/local/libexec/flux/flux-shell
2022-04-19T01:42:36.695310Z job-exec.debug[0]: using imp path /usr/local/libexec/flux/flux-imp
2022-04-19T01:42:36.748683Z broker.debug[0]: insmod heartbeat
2022-04-19T01:42:36.753886Z broker.info[0]: rc1.0: running /usr/local/etc/flux/rc1.d/02-cron
2022-04-19T01:42:36.856625Z broker.debug[0]: insmod sched-simple
2022-04-19T01:42:36.857629Z sched-simple.debug[0]: service_register
2022-04-19T01:42:36.858625Z sched-simple.debug[0]: resource update: {"resources":{"version":1,"execution":{"R_lite":[{"rank":"1-7","children":{"core":"0-3"}}],"starttime":0.0,"expiration":0.0,"nodelist":["picl[1-7]"],"properties":{"4gb":"2,5"}}},"up":""}
2022-04-19T01:42:36.859398Z job-manager.debug[0]: scheduler: hello
2022-04-19T01:42:36.859866Z job-manager.debug[0]: scheduler: ready limited
2022-04-19T01:42:36.860138Z sched-simple.debug[0]: ready: 0 of 28 cores: 
2022-04-19T01:42:37.376050Z broker.info[0]: rc1.0: /usr/local/etc/flux/rc1 Exited (rc=0) 2.9s
2022-04-19T01:42:37.376312Z broker.info[0]: rc1-success: init->quorum 2.90222s
2022-04-19T01:42:37.476974Z broker.debug[0]: groups: broker.online=0
2022-04-19T01:42:37.477244Z broker.info[0]: online: picl0 (ranks 0)
2022-04-19T01:42:37.477359Z broker.info[0]: quorum-full: quorum->run 0.101016s
2022-04-19T01:42:37.477496Z broker.debug[0]: rc2-none: ignored in run
2022-04-19T01:42:37.478852Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:10.725705Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:10.738474Z broker.info[0]: accepting connection from picl1 (rank 1) uuid 03b0ed7f-56ab-47cf-b08f-374f9e77883d status full
2022-04-19T02:07:10.848737Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:10.850098Z broker.info[0]: accepting connection from picl3 (rank 3) uuid e00982d8-eb3d-4a3b-9044-ccfc64b6ad8e status full
2022-04-19T02:07:11.429940Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:11.431422Z broker.info[0]: accepting connection from picl7 (rank 7) uuid e2165bbd-c161-4c62-951d-ccbcd72a2d7b status full
2022-04-19T02:07:11.772617Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:11.774095Z broker.info[0]: accepting connection from picl6 (rank 6) uuid 1678e2e4-7049-4894-8a55-82d4f73abc4b status full
2022-04-19T02:07:11.963362Z broker.debug[0]: groups: broker.online=0-1
2022-04-19T02:07:11.963983Z broker.info[0]: online: picl[0-1] (ranks 0-1)
2022-04-19T02:07:11.977636Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:11.978292Z sched-simple.debug[0]: resource update: {"up":"1"}
2022-04-19T02:07:12.090141Z broker.debug[0]: groups: broker.online=0-1,3
2022-04-19T02:07:12.090750Z broker.info[0]: online: picl[0-1,3] (ranks 0-1,3)
2022-04-19T02:07:12.097544Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:12.098170Z sched-simple.debug[0]: resource update: {"up":"3"}
2022-04-19T02:07:12.763911Z broker.debug[0]: groups: broker.online=0-1,3,7
2022-04-19T02:07:12.764617Z broker.info[0]: online: picl[0-1,3,7] (ranks 0-1,3,7)
2022-04-19T02:07:12.772435Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:12.773148Z sched-simple.debug[0]: resource update: {"up":"7"}
2022-04-19T02:07:12.826217Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:12.828705Z broker.info[0]: accepting connection from picl5 (rank 5) uuid d27e682b-faea-40f3-a458-e0513fd17bf9 status full
2022-04-19T02:07:12.930021Z broker.info[0]: overlay auth picl0 OK
2022-04-19T02:07:12.931718Z broker.debug[0]: groups: broker.online=0-1,3,6-7
2022-04-19T02:07:12.931889Z broker.info[0]: accepting connection from picl2 (rank 2) uuid 50ff70d8-8dec-49ce-87a2-4168d45251c0 status full
2022-04-19T02:07:12.932190Z broker.info[0]: online: picl[0-1,3,6-7] (ranks 0-1,3,6-7)
2022-04-19T02:07:12.937828Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:12.938321Z sched-simple.debug[0]: resource update: {"up":"6"}
2022-04-19T02:07:13.867667Z broker.debug[0]: groups: broker.online=0-1,3,5-7
2022-04-19T02:07:13.868421Z broker.info[0]: online: picl[0-1,3,5-7] (ranks 0-1,3,5-7)
2022-04-19T02:07:13.876089Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:13.876718Z sched-simple.debug[0]: resource update: {"up":"5"}
2022-04-19T02:07:14.049000Z broker.debug[0]: groups: broker.online=0-3,5-7
2022-04-19T02:07:14.049641Z broker.info[0]: online: picl[0-3,5-7] (ranks 0-3,5-7)
2022-04-19T02:07:14.056879Z resource.debug[0]: reslog_cb: online event posted
2022-04-19T02:07:14.057500Z sched-simple.debug[0]: resource update: {"up":"2"}
2022-04-19T02:07:40.733207Z sched-simple.debug[0]: req: 813193739894784: spec={7,7,1} duration=0.0
2022-04-19T02:07:40.733592Z sched-simple.debug[0]: alloc: 813193739894784: rank[1-7]/core[0-3]
2022-04-19T02:07:41.095768Z job-manager.err[0]: 813193739894784: prolog: stderr: picl4 (rank 4): flux-exec: Error: rank 4: /usr/local/libexec/flux/flux-imp: No route to host
2022-04-19T02:07:41.148820Z job-manager.err[0]: 813193739894784: prolog: stderr: F7MJDo9KGs: rank 4 failed prolog, draining
2022-04-19T02:07:41.152932Z resource.debug[0]: reslog_cb: drain event posted
2022-04-19T02:07:41.212189Z sched-simple.debug[0]: free: rank[1-7]/core[0-3]
```

**Comment by grondo** [comment]
Ugh, this is a dumb bug in the special case for handling `exclusive` node requests -- it completely ignores down nodes! :facepalm:

This patch will probably fix the issue: sort down nodes to the end of the list (hope I got that right), and abort the exclusive node search upon find the first non-idle or _down_ node.

```diff
diff --git a/src/common/librlist/rlist.c b/src/common/librlist/rlist.c
index 77540267a..fb08e05c1 100644
--- a/src/common/librlist/rlist.c
+++ b/src/common/librlist/rlist.c
@@ -1687,7 +1687,9 @@ static int by_used (const void *item1, const void *item2)
     int n;
     const struct rnode *x = item1;
     const struct rnode *y = item2;
-    if ((n = rnode_avail (y) - rnode_avail (x)) == 0)
+    if (x->up != y->up)
+        n = x->up ? -1 : 1;
+    else if ((n = rnode_avail (y) - rnode_avail (x)) == 0)
         n = by_rank (x, y);
     return n;
 }
@@ -1859,9 +1861,13 @@ static struct rlist *rlist_alloc_nnodes (struct rlist *rl,
         struct rnode *cpy;
         n = zlistx_first (rl->nodes);
         while (n && nleft) {
-            /*  We can abort after we find the first non-idle node.
+            /*
+             *  We can abort after we find the first non-idle or down node:
+             *
+             *  Note: by_used() sorts down nodes to back of list and
+             *   rnode_avail() returns 0 if node is not up
              */
-            if (idset_count (n->cores->avail) < idset_count (n->cores->ids))
+            if (rnode_avail (n) < rnode_count (n))
                 goto unwind;
 
             if (!(cpy = rnode_copy (n))
```

I was able to demonstrate this problem with something as simple as:
```console
$ src/cmd/flux start -s 2 'flux cron interval -c1 5s flux resource undrain 0 && flux resource drain 0 && flux mini run -N2 flux getattr rank'
```
The command should wait for 5s until the cron job fires, but currently the job runs right away as you've seen.

With the patch above the command works as expected. I'll post a PR tomorrow with some tests.

**Comment by grondo** [comment]
Should have been closed by #4292
